### PR TITLE
WIP: Add tests for HttpPostContext content-type

### PR DESCRIPTION
**Pull request by mlevesquedion**
I've added a couple tests which I think cover most of the functionality. I couldn't get the first way (header) to work.

I don't believe this is ready to merge, hence the "WIP" in the title. Mostly looking to see whether I'm going in the right direction or not.

Cheers!

### CODE REVIEW

**Comment by rybalkinsd** [comment]
hey @mlevesquedion !

You're on the right way. However, it's also important to check the case:

```kotlin
httpPost {
    host = 
    path =
    header {
        "Content-Type" to ...
    }
    body(type) {}
}
```
OkHttp has an [issue](https://github.com/square/okhttp/issues/2099) related to the topic. However header "Content-Type" might be ignored in okhttp (important to check). 

Also, it's important to add a note to `body { }` DSL about content type resolution in case of multi-declaration.

**Comment by mlevesquedion** [comment]
Hey there!

Thanks for your feedback! I'll add the missing tests and the note later today.

**Comment by mlevesquedion** [comment]
Here's an update. If I understood you correctly, this should do it.

Summary of my changes :
1. Get content type from header as default argument for body
2. Add tests covering remaining features, removed one test which was not necessary
3. Modified the README to show how to set the content type in the header and inform users about content type resolution

**Comment by rybalkinsd** [comment]
@mlevesquedion 
I might be not concrete enough with the details.
As mentioned in #62 task scope, I expected only tests and readme.md to be changed.
According to this I suggest to leave only related changes in this PR. 
/src changes better to provide in a separate PR. ( it would be much easier for release. Test and docs could be merged instantly, but sources will be merged in next release brunch)

Btw, did you check the behavior of okhttp3 for case when post request has both body with Content-Type and header with different type?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpContext.kt [hunk 1]
```diff
@@ -66,12 +66,16 @@ sealed class HttpContext(private val method: Method = GET) : IHttpContext {
...
+    }
+
 }
 
-open class HttpPostContext(method: Method = POST): HttpContext(method) {
+open class HttpPostContext(method: Method = POST) : HttpContext(method) {
     private var body: RequestBody = RequestBody.create(null, byteArrayOf())
 
-    fun body(contentType: String? = null, init: BodyContext.() -> RequestBody) {
+    fun body(contentType: String? = getHeaderContentType(), init: BodyContext.() -> RequestBody) {
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpContext.kt]: What if 
Body declared before header?

`httpPost { body { } header { } }`

File: src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpPostContextTest.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        val headerContentType = "text/css; charset=utf-8"
+        val customContentType = "text/html; charset=utf-8"
+        val context = HttpPostContext()
+        context.header { "Content-Type" to headerContentType }
+        context.body(customContentType) { string("content") }
+        assertEquals(customContentType, context.makeBody().contentType().toString())
+    }
+
+    @Test
+    fun `should override custom content type with type inside body`() {
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HttpPostContextTest.kt]: This test proves that changes in readme are not fully correct

File: README.md [hunk 3]
```diff
@@ -162,6 +162,25 @@ httpPost {
 }
 ```
 
+You may also specify a custom content type in the request's header :
+```kotlin
+httpPost {
+    header { "Content-Type" to "text/html" }
```
Comment (rybalkinsd) [inline README.md]: Not sure that we need to recommend this option.
It’s better to mention in body javadoc.

File: src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HeaderContext.kt [hunk 4]
```diff
@@ -18,4 +18,8 @@ class HeaderContext {
     }
 
     internal fun forEach(action: (k: String, v: Any) -> Unit) = map.forEach(action)
+
+    fun getContentType(): String? {
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HeaderContext.kt]: As I previously commented, suggest to move /src changes to separate PR

Comment (rybalkinsd) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/context/HeaderContext.kt]: Also looks like this function better to be private

File: README.md [hunk 5]
```diff
@@ -162,6 +162,25 @@ httpPost {
...
+    }
+}
+```
+
+
+##### Content type priority
+
+Content type is set according to the following priority levels (higher is prioritized) :
+
+1. Header : ```kotlin ... header { "Content-Type" to "$myContenType" } ... ```
```
Comment (rybalkinsd) [inline README.md]: Please double check the priority ordering

Comment (rybalkinsd) [inline README.md]: In case we move /src changes to separate PR, will header be first?

**Comment by mlevesquedion** [comment]
Hi @rybalkinsd,

Thank you for the comments! I haven't been able to look at this in the past few days, but I might have some time to look at it this weekend. In the meantime, if you want to go with a different approach, that is totally okay with me. Cheers!

**Comment by rybalkinsd** [comment]
@mlevesquedion Take your time! We're not urgent with this task. In my mind we can iteratively get to the production solution soon!

**Comment by doyaaaaaken** [comment]
This PR is stale because issue #62 is already resolved.
So, maybe we should close this PR.

**Pull request closed without merging**